[asr] OpenRouter Whisper 长语音触发 15 秒全局超时

### 现象

我在使用 ASR 时遇到一个和长语音相关的超时问题。

背景：

- 之前试过本地 Qwen ASR 1.7B，偶尔会有识别错误。
- 后来换过 Qwen 0.7B ASR，基本不会 timeout，但识别效果一般。
- 之后切到 OpenRouter 的 `openai/whisper-large-v3-turbo`，短一点的语音可以正常使用，但稍微长一点就会识别超时。

当前配置：

- ASR 供应商：OpenRouter Whisper
- Endpoint：`https://openrouter.ai/api/v1`
- Model：`openai/whisper-large-v3-turbo`
- API key 本身在其他请求里可用；短音频也能成功，所以不像是 key 或 endpoint 错误。

复现现象：

- 短语音可以正常转写。
- 稍微长一点的语音，录音能正常结束，但停止录音后卡在转写阶段，约 15 秒后失败。
- 日志中可以看到 recorder 已释放，然后 `whisper 全局超时 15 秒`。

相关日志片段：

```text
2026-06-20T09:31:20.775532Z [INFO] [coord] hotkey pressed (mode=Toggle, phase=Listening)
2026-06-20T09:31:20.860231Z [INFO] [recorder] cpal Stream dropped (mic released)
2026-06-20T09:31:36.561286Z [ERROR] [coord] whisper 全局超时 15 秒
2026-06-20T09:31:36.562472Z [INFO] [coord] hotkey released (mode=Toggle, phase=Idle)
```

我的疑问：

- 这个问题是否是因为 Whisper/OpenRouter 路径仍然使用固定的 `COORDINATOR_GLOBAL_TIMEOUT_SECS = 15`？
- OpenRouter 这类云端 batch ASR 会受网络抖动、排队、base64 body、分片数量影响，15 秒整体超时是不是太严格了？
- 如果网络稍微 shake 一下就 error，而且失败后音频没有保留/没有 retry 入口，用户很难重新跑这一段，只能重新录，很难受。

相关上下文：

- #353 / #434 似乎已经针对本地 Qwen ASR 做过长录音动态超时修复。
- #582 / #591 已经支持 OpenRouter ASR，并且 OpenRouter 长录音会按 30s 分片。
- 但从现象看，OpenRouter Whisper 的总转写阶段仍然会被 15 秒全局超时限制住。

### 影响

- 长一点的语音输入不稳定：短句可用，稍长就可能失败。
- 失败原因对用户不直观，看起来像 provider/key 问题，但实际可能是客户端 timeout。
- 失败后如果音频没有保留或不能从失败历史重试，用户只能重新录音。
- 使用 OpenRouter Whisper Large V3 Turbo 作为高质量低成本 ASR 的体验会明显受影响。

### 建议接受标准

- [ ] Whisper-compatible ASR，尤其是 OpenRouter Whisper 路径，不再使用固定 15 秒整体超时。
- [ ] OpenRouter Whisper 根据音频长度、分片数量或 provider 类型使用动态 timeout，例如 `max(30s, audio_seconds * factor + margin)`，或者至少允许长音频有更合理的总预算。
- [ ] 日志中记录本次音频时长、分片数量、provider、timeout 秒数，方便判断是 provider 慢还是客户端超时。
- [ ] OpenRouter 分片请求可以承受短暂网络抖动；如果失败，错误信息区分 provider HTTP error、network error、client timeout。
- [ ] 转写失败时提供可用的重试路径：保留本次音频/PCM 到失败历史，或者至少在 debug/设置中允许保留并重新转写，避免用户必须重新录音。
- [ ] 不回退 #591 已修复的 OpenRouter JSON + base64 请求格式。

### TODO / 不确定项

- 我还没有确认具体 release 版本号；从 UI 看已经有 OpenRouter Whisper preset，应该是包含 #591 的版本。
- 我不确定当前失败后音频是否仍在内存中短暂存在，但 UI/历史里没有明显的重新转写入口。
- 需要维护者确认 OpenRouter Whisper 路径当前是否确实被 `COORDINATOR_GLOBAL_TIMEOUT_SECS = 15` 包住，以及是否可以复用本地 Qwen ASR 的动态 timeout 思路。

--- a/openless-all/app/src-tauri/src/asr/whisper.rs
+++ b/openless-all/app/src-tauri/src/asr/whisper.rs
@@ -87,6 +87,12 @@ impl WhisperBatchASR {
     /// Whisper transcriptions endpoint.
     ///
     /// 失败时**保留** PCM buffer，让上层有机会重试或在历史中至少留一个失败记录；
+    /// 当前缓冲音频时长（毫秒）。Coordinator 在 transcribe() 调用前读取，
+    /// 用于计算 Whisper / OpenRouter 的动态超时。不消费缓冲。
+    pub fn buffer_duration_ms(&self) -> u64 {
+        pcm_duration_ms(&self.buffer.lock())
+    }
+
     /// 之前的实现一进函数就 `mem::take` 把 buffer 清空，凭证错或网络中断都会
     /// 让用户的录音直接消失。
     pub async fn transcribe(&self) -> Result<RawTranscript> {

--- a/openless-all/app/src-tauri/src/coordinator.rs
+++ b/openless-all/app/src-tauri/src/coordinator.rs
@@ -2897,9 +2897,9 @@ const CAPSULE_AUTO_HIDE_DELAY_MS: u64 = 2000;
 const POST_SESSION_COOLDOWN_MS: u64 = 600;
 
 /// Coordinator 全局超时保护：防止 ASR await_final_result() 永远挂起。
-/// 设置为 15 秒（比 ASR 的 12 秒 FINAL_RESULT_TIMEOUT 稍长），
-/// 只在 ASR 超时机制失效时作为最后的防线触发。
-const COORDINATOR_GLOBAL_TIMEOUT_SECS: u64 = 15;
+/// 设置为 30 秒，为云端 batch ASR（OpenRouter Whisper 等）提供足够的
+/// 网络超时预算；只在 ASR 自身超时机制失效时作为最后的防线触发。
+const COORDINATOR_GLOBAL_TIMEOUT_SECS: u64 = 30;
 
 #[cfg(target_os = "windows")]
 fn foundry_audio_transcribe_timeout_duration() -> std::time::Duration {
@@ -2913,6 +2913,17 @@ fn foundry_audio_transcribe_timeout_duration() -> std::time::Duration {
 fn local_qwen_transcribe_timeout(audio_secs: f64) -> std::time::Duration {
     let secs = ((audio_secs * 0.6).ceil() as u64)
         .saturating_add(10)
+        .max(COORDINATOR_GLOBAL_TIMEOUT_SECS);
+    std::time::Duration::from_secs(secs)
+}
+
+/// Whisper / OpenRouter 云端 batch ASR 的动态转写超时。OpenRouter 按 30s
+/// 分片，每片是一次 HTTP round-trip；网络抖动、排队、base64 body 都会
+/// 拉长耗时。公式 max(30, ceil(audio_s × 0.5) + 20)：30s 是全局兜底；
+/// 长录音按音频长度的 0.5 倍 + 20s 余量，覆盖多分片串行请求 + 网络波动。
+fn whisper_transcribe_timeout(audio_secs: f64) -> std::time::Duration {
+    let secs = ((audio_secs * 0.5).ceil() as u64)
+        .saturating_add(20)
         .max(COORDINATOR_GLOBAL_TIMEOUT_SECS);
     std::time::Duration::from_secs(secs)
 }

--- a/openless-all/app/src-tauri/src/coordinator.rs
+++ b/openless-all/app/src-tauri/src/coordinator.rs
@@ -2326,7 +2326,7 @@ mod tests {
 
     #[test]
     fn local_qwen_timeout_floors_at_global_timeout_for_short_audio() {
-        // 5s 录音：5 × 0.6 = 3, +10 = 13, max(15) = 15。短录音保留 15s 兜底。
+        // 5s 录音：5 × 0.6 = 3, +10 = 13, max(30) = 30。短录音兜底。
         assert_eq!(
             local_qwen_transcribe_timeout(5.0),
             std::time::Duration::from_secs(COORDINATOR_GLOBAL_TIMEOUT_SECS)
@@ -2344,19 +2344,47 @@ mod tests {
 
     #[test]
     fn local_qwen_timeout_ceils_partial_seconds() {
-        // 10.1s 录音：10.1 × 0.6 = 6.06, ceil = 7, +10 = 17, max(15) = 17。
+        // 10.1s 录音：10.1 × 0.6 = 6.06, ceil = 7, +10 = 17, max(30) = 30。
+        // COORDINATOR_GLOBAL_TIMEOUT_SECS 提升到 30 后，短音频统一被兜底值覆盖。
         assert_eq!(
             local_qwen_transcribe_timeout(10.1),
-            std::time::Duration::from_secs(17)
+            std::time::Duration::from_secs(COORDINATOR_GLOBAL_TIMEOUT_SECS)
         );
     }
 
     #[test]
     fn local_qwen_timeout_handles_zero_duration() {
-        // 0 时长（空 buffer 边界）：0 × 0.6 = 0, +10 = 10, max(15) = 15。
+        // 0 时长（空 buffer 边界）：0 × 0.6 = 0, +10 = 10, max(30) = 30。
         assert_eq!(
             local_qwen_transcribe_timeout(0.0),
             std::time::Duration::from_secs(COORDINATOR_GLOBAL_TIMEOUT_SECS)
+        );
+    }
+
+    #[test]
+    fn whisper_timeout_floors_at_global_timeout_for_short_audio() {
+        // 10s 录音：10 × 0.5 = 5, +20 = 25, max(30) = 30。短音频兜底。
+        assert_eq!(
+            whisper_transcribe_timeout(10.0),
+            std::time::Duration::from_secs(COORDINATOR_GLOBAL_TIMEOUT_SECS)
+        );
+    }
+
+    #[test]
+    fn whisper_timeout_scales_with_audio_duration() {
+        // 60s 录音：60 × 0.5 = 30, +20 = 50。覆盖多分片 HTTP 请求。
+        assert_eq!(
+            whisper_transcribe_timeout(60.0),
+            std::time::Duration::from_secs(50)
+        );
+    }
+
+    #[test]
+    fn whisper_timeout_ceils_partial_seconds() {
+        // 45.3s 录音：45.3 × 0.5 = 22.65, ceil = 23, +20 = 43, max(30) = 43。
+        assert_eq!(
+            whisper_transcribe_timeout(45.3),
+            std::time::Duration::from_secs(43)
         );
     }
 

--- a/openless-all/app/src-tauri/src/coordinator/dictation.rs
+++ b/openless-all/app/src-tauri/src/coordinator/dictation.rs
@@ -1691,8 +1691,15 @@ pub(super) async fn end_session(inner: &Arc<Inner>) -> Result<(), String> {
         }
         ActiveAsr::Whisper(w) => {
             debug_assert!(uses_global_timeout);
-            // Whisper 也添加类似的超时保护
-            let timeout_duration = std::time::Duration::from_secs(COORDINATOR_GLOBAL_TIMEOUT_SECS);
+            // Whisper / OpenRouter 动态超时：音频越长、分片越多，给更多
+            // HTTP round-trip 预算。公式见 `whisper_transcribe_timeout`。
+            let audio_secs = (w.buffer_duration_ms() as f64) / 1000.0;
+            let timeout_duration = whisper_transcribe_timeout(audio_secs);
+            log::info!(
+                "[coord] Whisper transcribe: audio={:.2}s timeout={}s",
+                audio_secs,
+                timeout_duration.as_secs()
+            );
             match tokio::time::timeout(timeout_duration, w.transcribe()).await {
                 Ok(Ok(r)) => r,
                 Ok(Err(e)) => {
@@ -1712,8 +1719,9 @@ pub(super) async fn end_session(inner: &Arc<Inner>) -> Result<(), String> {
                 }
                 Err(_) => {
                     log::error!(
-                        "[coord] whisper 全局超时 {} 秒",
-                        COORDINATOR_GLOBAL_TIMEOUT_SECS
+                        "[coord] Whisper 动态超时 {}s（音频 {:.2}s）",
+                        timeout_duration.as_secs(),
+                        audio_secs
                     );
                     emit_capsule(
                         inner,

